### PR TITLE
Implement vectorized FMA of Double and Float Vectors

### PR DESCRIPTION
**Pull request by midronij**
Implements codegen for vectorized FMA of Double and Float Vectors 

Note: This PR should not be merged until the IL opcode for vector FMA is implemented

### CODE REVIEW

**Comment by midronij** [comment]
@gita-omr @IBMJimmyk could you please review when you have a chance?

File: compiler/p/codegen/OMRTreeEvaluator.cpp [hunk 1]
```diff
@@ -2759,8 +2759,7 @@ TR::Register *OMR::Power::TreeEvaluator::inlineVectorBitSelectOp(TR::Node *node,
 
    TR::Register *resReg;
 
-   if (op != TR::InstOpCode::xvmaddadp &&
-       op != TR::InstOpCode::xvmsubadp &&
+   if (op != TR::InstOpCode::xvmsubadp &&
```
Comment (aviansie-ben) [inline compiler/p/codegen/OMRTreeEvaluator.cpp:2762]: Can you please clarify what the reason behind this change is? It seems unrelated to the rest of this change AFAICT.

Comment (midronij) [reply compiler/p/codegen/OMRTreeEvaluator.cpp:2762]: Ah, I think that's a leftover from an earlier draft that I forgot to clean up. Thanks for catching that!

**Comment by midronij** [comment]
Now that the PRs for the new vector IL opcodes are merged (https://github.com/eclipse/omr/pull/6458 and https://github.com/eclipse-openj9/openj9/pull/14855), this one should be ready to go. @gita-omr @IBMJimmyk @aviansie-ben could you please review when you have a chance?

File: compiler/p/codegen/OMRTreeEvaluator.cpp [hunk 2]
```diff
@@ -3850,7 +3850,56 @@ TR::Register *OMR::Power::TreeEvaluator::vdsqrtEvaluator(TR::Node *node, TR::Cod
 
 TR::Register* OMR::Power::TreeEvaluator::vfmaEvaluator(TR::Node *node, TR::CodeGenerator *cg)
    {
-   return TR::TreeEvaluator::unImpOpEvaluator(node, cg);
+   TR::DataType type = node->getDataType().getVectorElementType();
+   if (type != TR::Float && type != TR::Double)
+   {
+      TR_ASSERT(false, "unsupported vector type %s\n", type.toString());
+      return NULL;
+   }
```
Comment (IBMJimmyk) [inline compiler/p/codegen/OMRTreeEvaluator.cpp:3858]: Brackets need to be fixed here.

File: compiler/p/codegen/OMRTreeEvaluator.cpp [hunk 3]
```diff
@@ -3850,7 +3850,56 @@ TR::Register *OMR::Power::TreeEvaluator::vdsqrtEvaluator(TR::Node *node, TR::Cod
...
+
+   TR::InstOpCode::Mnemonic op;
+
+   TR::Node *firstChild  = node->getFirstChild();
+   TR::Node *secondChild = node->getSecondChild();
+   TR::Node *thirdChild  = node->getThirdChild();
+
+   TR::Register *thirdReg  = cg->evaluate(thirdChild);
+   TR::Register *firstReg  = cg->evaluate(firstChild);
+   TR::Register *secondReg = cg->evaluate(secondChild);
```
Comment (IBMJimmyk) [inline compiler/p/codegen/OMRTreeEvaluator.cpp:3868]: This should be in order unless there's a reason they are not?

File: compiler/p/codegen/OMRTreeEvaluator.cpp [hunk 4]
```diff
@@ -3850,7 +3850,56 @@ TR::Register *OMR::Power::TreeEvaluator::vdsqrtEvaluator(TR::Node *node, TR::Cod
...
+   {
+      TR_ASSERT(false, "unsupported vector type %s\n", type.toString());
+      return NULL;
+   }
+
+   TR::InstOpCode::Mnemonic op;
+
+   TR::Node *firstChild  = node->getFirstChild();
+   TR::Node *secondChild = node->getSecondChild();
+   TR::Node *thirdChild  = node->getThirdChild();
```
Comment (IBMJimmyk) [inline compiler/p/codegen/OMRTreeEvaluator.cpp:3864]: I think a comment here indicating which child refers to which part of the fma would be helpful here.

File: compiler/p/codegen/OMRTreeEvaluator.cpp [hunk 5]
```diff
@@ -3850,7 +3850,56 @@ TR::Register *OMR::Power::TreeEvaluator::vdsqrtEvaluator(TR::Node *node, TR::Cod
...
+
+   TR::Register *thirdReg  = cg->evaluate(thirdChild);
+   TR::Register *firstReg  = cg->evaluate(firstChild);
+   TR::Register *secondReg = cg->evaluate(secondChild);
+
+   TR::Register *resReg;
+
+   if (cg->canClobberNodesRegister(thirdChild))
```
Comment (IBMJimmyk) [inline compiler/p/codegen/OMRTreeEvaluator.cpp:3872]: A comment around here explaining that the code is searching for a register to clobber because the instruction will clobber one of its operands with the result would be nice here.

File: compiler/p/codegen/OMRTreeEvaluator.cpp [hunk 6]
```diff
@@ -3850,7 +3850,56 @@ TR::Register *OMR::Power::TreeEvaluator::vdsqrtEvaluator(TR::Node *node, TR::Cod
...
+      }
+   else if (cg->canClobberNodesRegister(secondChild))
+      {
+      resReg = secondReg;
+      op = (type == TR::Float) ? TR::InstOpCode::xvmaddmsp : TR::InstOpCode::xvmaddmdp;
+      generateTrg1Src2Instruction(cg, op, node, secondReg, firstReg, thirdReg);
+      }
+   else
+      {
+      resReg = cg->allocateRegister(TR_VSX_VECTOR);
```
Comment (IBMJimmyk) [inline compiler/p/codegen/OMRTreeEvaluator.cpp:3892]: A comment here saying that all of the operand registers are still needed later so the code needs to copy one of the operands for this to work.

**Review by IBMJimmyk — CHANGES_REQUESTED** [review]
The code looks good. My requests are largely for more comments.

**Review by IBMJimmyk — APPROVED** [review]
Looks good to me.

**Review by aviansie-ben — APPROVED** [review]
LGTM

**Comment by aviansie-ben** [comment]
Jenkins build plinux,aix
